text/0000-pi-types.md

```diff
@@ -0,0 +1,306 @@
+      Γ ⊢ c: A
+      Γ ⊢ x: A
+      Γ ⊢ a: T c
+      Γ ⊢ a: T x
+      ────────────
+      Γ ⊢ c = x
+
+So, if two types share constructor by some Π-constructor, share a value, their
+value parameter is equal. Take `a: [u8; 4]` as an example. If we have some
+unknown variable `x`, such that `a: [u8; x]`, we can infer that `x = 4`.
+
+This allows us to infer e.g. array length parameters in functions:
+
+```rust
+// [T; N] is a constructor, T → usize → 𝓤 (parameterize over T and you get A → 𝓤).
+fn foo<const n: usize, const l: [u32; n]>() -> [u32; n] {
+    // ^ note how l depends on n.
+    l
+}
+
+// We know n from the length of the array.
+let l = foo::<_, [1, 2, 3, 4, 5, 6]>();
+//            ^   ^^^^^^^^^^^^^^^^
+```
+
+## Structural equality
+
+Structural equality plays a key role in type checking of dependent types.
+
+Structural equality, in this case, is defined as an equivalence relation, which
+allows substitution without changing semantics.
+
+Any constant parameter must have the `structural_match` property as defined in
+[RFC #1445](https://github.com/rust-lang/rfcs/pull/1445). This property, added
+through the `#[structural_match]` attribute, essentially states that the `Eq`
+implementation is structural.
+
+Without this form of equality, substitution wouldn't be possible, and thus
+typechecking an arbitrarily value-depending type constructor would not be
+possible.
+
+## Parsing
+
+Introducing expr subgrammar in type position isn't possible without setting
+some restrictions to the possible expressions.
+
+In this case, we put restrictions to arithmetic and bitwise operations (`+-/!^`
+etc.) and function calls (`myconstfn(a, b, c, ...)` with `a, b, c, ...` being
+`const-expr` fragments).
+
+Additionally, we allow parenthesizes, which can contain any general const-expr
+fragment.
+
+# How we teach this
+
+This RFC aims to keep a "symmetric" syntax to the current construct, giving an
+intuitive behavior, however there are multiple things that are worth explaining
+and/or clearing up:
+
+**What are dependent types?**
+
+Dependent types are types, which _depend_ on values, instead of types. For
+example, [T; 3], is dependent since it depends on the value, `3`, for
+constructing the type. Dependent types, in a sense, are similar to normal
+generics, where types can depend on other types (e.g. `Vec<T>`), whereas
+dependent types depend on values.
+
+**How does this differ from other languages' implementations of dependent types?**
+
+Various other languages have dependent type systems. Strictly speaking, all
+that is required for a dependent type system is value-to-type constructors,
+although some languages (coq, agda, etc.) goes a step further and remove the
+boundary between value and type. Unfortunately, as cool as it sounds, it has
+some severe disadvantages: most importantly, the type checking becomes
+undecidable. Often you would need some form of theorem prover to type check
+the program, and those have their limitations too.
```

Comment:
I just want to say that the part about undecidability here is very misleading.

To be clear, type checking in Agda and Coq is *decidable* in general. Undecidability potentially comes into play from non-termination of some program fragment during evaluation that occurs while type checking. But by default, both Agda and Coq require that programs pass the termination check *before* they can be used. So that source of undecidability cannot arise unless termination checking is explicitly disabled.

The case may be different for Idris since it does not enforce termination by default as far as I know.

However, even in the case where there *is* undecidability, it's also misleading to bring it up as a serious disadvantage in the context of Rust. The fact of the matter is that type checking in Rust is *already undecidable* due to the expressive power of associated types. For instance, you can already encode a full lambda calculus interpreter in the type system.